test/test_filter.py

```diff
@@ -120,136 +115,197 @@ def test_filter_coeff(actx_factory, filter_order, order, dim):
-    filtered_field = filter_modally(dcoll, cutoff, frfunc, field)
-    soln_resid = field - filtered_field
-    max_errors = [actx.to_numpy(op.norm(dcoll, v, np.inf)) for v in soln_resid]
-    logger.info(f"Field = {field}")
-    logger.info(f"Filtered = {filtered_field}")
-    logger.info(f"Max Errors (poly) = {max_errors}")
-    assert np.max(max_errors) < tol
-
-    # Any order > cutoff fields should have higher modes attenuated
-    threshold = 1e-3
-    tol = 1e-1
-    if do_viz is True:
-        from grudge.shortcuts import make_visualizer
-        vis = make_visualizer(dcoll, order)
-
-    from grudge.dof_desc import DD_VOLUME_ALL, DD_VOLUME_ALL_MODAL
-
-    modal_map = dcoll.connection_from_dds(DD_VOLUME_ALL, DD_VOLUME_ALL_MODAL)
-
-    for field_order in range(cutoff+1, cutoff+4):
-        coeff = [1.0 / (i + 1) for i in range(field_order+1)]
-        field = polyfn(coeff=coeff)
-        filtered_field = filter_modally(dcoll, cutoff, frfunc, field)
-
-        unfiltered_spectrum = modal_map(field)
-        filtered_spectrum = modal_map(filtered_field)
-        if do_viz is True:
-            spectrum_resid = unfiltered_spectrum - filtered_spectrum
-            io_fields = [
-                ("unfiltered", field),
-                ("filtered", filtered_field),
-                ("unfiltered_spectrum", unfiltered_spectrum),
-                ("filtered_spectrum", filtered_spectrum),
-                ("residual", spectrum_resid)
-            ]
-            vis.write_vtk_file(f"filter_test_{field_order}.vtu", io_fields)
-        field_resid = unfiltered_spectrum - filtered_spectrum
-        max_errors = [actx.to_numpy(op.norm(dcoll, v, np.inf)) for v in field_resid]
-        # fields should be different, but not too different
-        assert (tol > np.max(max_errors) > threshold)
+    # ISO fields are for hand-testing, please don't remove
+    # iso_fields = []  # f(x) = x**order
+    # iso_cutoff = 2
+    fields = []  # f(x) = a0 + a1*x + a2*x*x + ....
+    numfields = len(test_field_orders)
+
+    for field_order in test_field_orders:
+        # iso_field_coeff = [1.0 if i == field_order else 0.0
+        #                   for i in range(field_order+1)]
+        field_coeff = [1.0 / (i + 1) for i in range(field_order+1)]
+        fields.append(poly_func(field_coeff))
+        # iso_fields.append(poly_func(iso_field_coeff))
+
+    unfiltered_fields = make_obj_array(fields)
+    # unfiltered_iso_fields = make_obj_array(iso_fields)
+    unfiltered_spectra = modal_map(unfiltered_fields)
+    spectra_numbers = actx.to_numpy(unfiltered_spectra)
+
+    # print(f"{unfiltered_spectra=}")
+    # print(f"{spectra_numbers.shape=}")
+    # print(f"{spectra_numbers=}")
+
+    element_spectra = make_obj_array([
+        [np.array([0. for i in range(element_order+1)])
+         for j in range(numfields)] for k in range(numelem)])
+
+    total_power = [np.array([0. for _ in range(element_order+1)])
+                   for e in range(numfields)]
```

Comment:
It looks like this and the code below could be simplified by using multidimensional numpy arrays instead of these nested lists of arrays. At a glance I can't tell if there's anything getting in the way of doing that or not. For example, if you could find a way to compute `element_spectra` with numpy operations, then I think `total_power` would just be something like `sum(element_spectra, axis=0)`.

Reply:
I took a stab at this in [8a2bb98](https://github.com/illinois-ceesd/mirgecom/pull/837/commits/8a2bb98759b4e2338e3b97f7d45e373ef1e9e2a6). Feel free to revert/edit as desired.

Reply:
Outstanding!  Thanks!

Reply:
Please see what you think of this - just moving your changes off into a utility that we can use outside of testing.
([71f1708](https://github.com/illinois-ceesd/mirgecom/pull/837/commits/71f17082ab63973782b097c5a93a9b007c8e162d))